File: 2015/21/fight.hs
import Control.Arrow ((&&&))
import Data.List (subsequences)

data Item = Item {
    cost :: Int,
    damage :: Int,
    protection :: Int
} deriving (Show)

data Fighter = Fighter {
    hp :: Int,
    attack :: Int,
    defence :: Int
} deriving (Show)

genItems :: [(Int, Int, Int)] -> [Item]
genItems = map (\(c, d, a) -> Item { cost=c, damage=d, protection=a })

bossInitial :: Fighter
bossInitial = Fighter { hp = 104, attack = 8, defence = 1 }

player :: [Item] -> Fighter
player = foldl (\p x -> p {attack = attack p + damage x,
                           defence = defence p + protection x}) playerInitial
    where playerInitial = Fighter { hp = 100, attack = 0, defence = 0 }

choices :: [[Item]]
choices = base ++ [b ++ r | b <- base, r <- rfilt 1]
               ++ [b ++ r | b <- base, r <- rfilt 2]
    where rfilt n = filter ((== n) . length) (subsequences rings)
          base = [[w] | w <- weapons] ++ [[w,a] | w <- weapons, a <- armor]
          weapons = genItems [(8, 4, 0), (10, 5, 0), (25, 6, 0), (40, 7, 0), (74, 8, 0)]
          armor   = genItems [(13, 0, 1), (31, 0, 2), (53, 0, 3), (75, 0, 4), (102, 0, 5)]
          rings   = genItems [(25, 1, 0), (50, 2, 0), (100, 3, 0),
                               (20, 0, 1), (40, 0, 2), (80, 0, 3)]

fight :: [Item] -> Bool
fight xs = hp (fight' (player xs) bossInitial False) > 0
    where fight' a b flipped
            | hp a <= 0 || hp b <= 0 = if flipped then b else a
            | otherwise = fight' (strike a b) a (not flipped)
          strike a b = b { hp = hp b - max 1 (attack a - defence b) }

search :: (Int, Int)
search = (part1, part2)
    where results = map (fight &&& totalCost) choices
          part1 = minimum [c | (won, c) <- results, won]
          part2 = maximum [c | (won, c) <- results, not won]
          totalCost xs = sum [cost x | x <- xs]

main :: IO ()
main = do
    let (part1, part2) = search
    putStrLn $ "Part 1: " ++ show part1
    putStrLn $ "Part 2: " ++ show part2

File: 2015/21/fight.py
weapons = [(8, 4), (10, 5), (25, 6), (40, 7), (74, 8)]
armor = [(13, 1), (31, 2), (53, 3), (75, 4), (102, 5)]
rings = [(25, 1, 0), (50, 2, 0), (100, 3, 0), (20, 0, 1), (40, 0, 2), (80, 0, 3)]


def fight(boss, weapon, armor, rings):
    player_damage = weapon[1] + sum([x[1] for x in rings])
    player_defence = armor[1] + sum([x[2] for x in rings])
    p = [[100, player_damage, player_defence], boss[:]]

    hp, damage, defence = 0, 1, 2

    turn = 0
    while p[0][hp] > 0 and p[1][hp] > 0:
        cur, other = turn % 2, (turn+1) % 2
        inflicted = p[cur][damage] - p[other][defence]
        if inflicted < 1:
            inflicted = 1
        p[other][hp] -= inflicted
        turn += 1
    winner = (turn+1) % 2

    cost = weapon[0] + armor[0] + sum([x[0] for x in rings])
    return winner == 0, cost

def check_win(best, r):
    won, cost = r
    if won:
        if cost < best:
            return cost
    return best

def check_loss(worst, r):
    won, cost = r
    if not won:
        if cost > worst:
            return cost
    return worst

def check(best, worst, r):
    return check_win(best, r), check_loss(worst, r)

def search(boss):
    results = []
    best = 1000
    worst = 0

    for weapon in weapons:
        for arm in [(0,0)] + armor:
            # 0 rings
            best, worst = check(best, worst, fight(boss, weapon, arm, []))

            # 1 ring
            for r in rings:
                best, worst = check(best, worst, fight(boss, weapon, arm, [r]))

            # 2 rings
            for r1 in range(len(rings)-1):
                for r2 in range(r1, len(rings)):
                    best, worst = check(best, worst, fight(boss, weapon, arm, [rings[r1], rings[r2]]))
    return best, worst

boss = [104, 8, 1]
best, worst = search(boss)
print 'Part 1:', best
print 'Part 2:', worst
